[Remote Store] Add segment transfer timeout dynamic setting

linuxpi opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
- Segment uploads for Remote Store happen without any timeout and can hold the thread on latch.await() for ever if there is an error which is swallowed silently by the code.
- Since the future is never completed, the shard is stuck and is never able to retry the upload
- This leads to remote time lag to increase and never come down until we restart the process
- This PR adds the timeout to segment uploads to recover automatically from such situations after the timeout.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/13783
<!-- List any other related issues here -->

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [ ] ~API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).~
- [ ] ~Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))~
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose) https://github.com/opensearch-project/documentation-website/issues/7227

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

skumawat2025 commented on server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:
```diff
@@ -286,7 +291,12 @@ public void onFailure(Exception e) {
+                        remoteStoreSettings.getClusterRemoteSegmentTransferTimeout().millis(),
+                        TimeUnit.MILLISECONDS
+                    ) == false) {
+                        throw new InterruptedException("Timeout while waiting for remote segment transfer to complete");
```
1. Just a thought if InterruptedException is the right exception here? As we are not actually interrupting the thread here, it's getting timed-out. 
2. Should we handle `InterruptedException` before we hit this timeout limit?

linuxpi replied:
1. Agree, i don't think we have a generic exception for segment upload failures. creating one!
2. `InterruptedException` would be handled by existing catch blocks and as of now i dont think we need to do any special handling in case of `InterruptedException`, a retry would be scheduled automatically.

skumawat2025 commented on server/src/main/java/org/opensearch/indices/RemoteStoreSettings.java:
```diff
@@ -105,9 +105,21 @@ public class RemoteStoreSettings {
+     */
+    public static final Setting<TimeValue> CLUSTER_REMOTE_SEGMENT_TRANSFER_TIMEOUT_SETTING = Setting.timeSetting(
+        "cluster.remote_store.segment.transfer_timeout",
+        TimeValue.timeValueHours(3),
```
Curious to know how we decided this time-out limit?

Bukhtawar replied:
We can keep it lower I feel, maybe 30 mins

skumawat2025 commented on server/src/test/java/org/opensearch/index/shard/RemoteStoreRefreshListenerTests.java:
```diff
@@ -638,7 +649,12 @@ private Tuple<RemoteStoreRefreshListener, RemoteStoreStatsTrackerFactory> mockIn
+            tracker,
+            DefaultRemoteStoreSettings.INSTANCE
+        );
         refreshListener.afterRefresh(true);
```
Can we have a test for the changes where we are hitting the time-limit?

linuxpi replied:
yes, working on adding tests

sachinpkale commented on server/src/main/java/org/opensearch/index/shard/SegmentUploadFailedException.java:
```diff
@@ -0,0 +1,26 @@
+     *
+     * @param message error message
+     */
+    public SegmentUploadFailedException(String message) {
```
Add `@opensearch.internal`

sachinpkale commented:
Please add changelog entry, fix the gradle build and complete the check-list in PR description.

linuxpi commented:
> Please add changelog entry, fix the gradle build and complete the check-list in PR description.

Thanks @sachinpkale for approving. I've added the necessary details and addressed your comment.
